Address book encryption

str4d opened the pull request:
Blocked on changes to the SDK to expose ZIP 32 Arbitrary key derivation.

> **Note**
>This code review checklist is intended to serve as a starting point for the author and reviewer, although it may not be appropriate for all types of changes (e.g. fixing a spelling typo in documentation).  For more in-depth discussion of how we think about code review, please see [Code Review Guidelines](../blob/main/docs/CODE_REVIEW_GUIDELINES.md).

# Author
<!-- NOTE: Do not modify these when initially opening the pull request.  This is a checklist template that you tick off AFTER the pull request is created. -->

- [x] **Self-review** your own code in GitHub's web interface[^1]
- [ ] Add **automated tests** as appropriate
- [ ] Update the [**manual tests**](../blob/main/docs/testing/manual_testing)[^2] as appropriate
- [ ] Check the **code coverage**[^3] report for the automated tests
- [x] Update **documentation** as appropriate (e.g [**README.md**](../blob/main/README.md), [**Architecture.md**](../blob/main/docs/Architecture.md), [**CHANGELOG.md**](../blob/main/CHANGELOG.md), etc.)
- [x] **Run the app** and try the changes
- [x] Pull in the latest changes from the **main** branch and **squash** your commits before assigning a reviewer[^4]

> **Note**
> It is good practice to provide before and after UI  **screenshots** in the description of this PR. This is only applicable for changes that modify the UI.

# Reviewer

- [ ] Check the code with the [Code Review Guidelines](../blob/main/docs/CODE_REVIEW_GUIDELINES.md) **checklist**
- [ ] Perform an **ad hoc review**[^5]
- [ ] Review the **automated tests**
- [ ] Review the **manual tests**
- [ ] Review the **documentation**, [**README.md**](../blob/main/README.md), [**Architecture.md**](../blob/main/docs/Architecture.md), etc. as appropriate
- [ ] **Run the app** and try the changes[^6]

[^1]: _Code often looks different when reviewing the diff in a browser, making it easier to spot potential bugs._
[^2]: _While we aim for automated testing of the application, some aspects require manual testing. If you had to manually test something during development of this pull request, write those steps down._
[^3]: _While we are not looking for perfect coverage, the tool can point out potential cases that have been missed. Code coverage can be generated with: `./gradlew check` for Kotlin modules and `./gradlew connectedCheck -PIS_ANDROID_INSTRUMENTATION_TEST_COVERAGE_ENABLED=true` for Android modules._
[^4]: _Having your code up to date and squashed will make it easier for others to review. Use best judgement when squashing commits, as some changes (such as refactoring) might be easier to review as a separate commit._
[^5]: _In addition to a first pass using the code review guidelines, do a second pass using your best judgement and experience which may identify additional questions or comments. Research shows that code review is most effective when done in multiple passes, where reviewers look for different things through each pass._
[^6]: _While the CI server runs the app to look for build failures or crashes, humans running the app are more likely to notice unexpected log messages, UI inconsistencies, or bad output data. Perform this step last, after verifying the code changes are safe to run locally._

nuttycom commented on ui-lib/src/main/java/co/electriccoin/zcash/ui/common/provider/AddressBookProvider.kt:
```diff
@@ -24,18 +33,71 @@ class AddressBookProviderImpl : AddressBookProvider {
+        val cipher = ChaCha20Poly1305.create(key)
+        val ciphertext = cipher.encrypt(plaintext, null)
+
+        outputStream.buffered().use {
+            it.write(salt)
+            it.write(ciphertext)
+        }
+    }
```
Can this please have a version byte prepended to it? Or is the version going to be stored somewhere else?

str4d replied:
There's a version byte in the plaintext, and my original understanding was that the ciphertext was going to be stored in a way that handled external versioning. But the wallet side has changed somewhat and I no longer know how the file is intended to be managed from that perspective.

daira replied:
There are two ways of doing versioning for the encryption, as opposed to the encrypted data:
a) derive keys with a domain separator (currently `"ZashiAddressBookEncryptionV1"`) and trial-decrypt with each such separator;
b) prepend a separate encryption version byte;
c) rely on external context.

I dislike c) but have no strong opinion between a) and b).

Milan-Cerovsky replied:
we currently do b)

daira commented on ui-lib/src/main/java/co/electriccoin/zcash/ui/common/provider/AddressBookProvider.kt:
```diff
@@ -23,19 +34,74 @@ class AddressBookProviderImpl : AddressBookProvider {
+
+        // Tink encodes the ciphertext as `nonce || ciphertext || tag`.
+        val cipher = ChaCha20Poly1305.create(key)
+        val ciphertext = cipher.encrypt(plaintext, null)
```
```suggestion
        val ciphertext = cipher.encrypt(plaintext, null)
        if (ciphertext.size() != 12 + plaintext.size() + 16) {
            throw AssertionError(
                "unexpected ChaCha20Poly1305 ciphertext length ${ciphertext.size()} bytes" +
                    " for plaintext length ${plaintext.size()} bytes")
        }
```
(I do not use Kotlin's [assert](https://kotlinlang.org/api/latest/jvm/stdlib/kotlin/assert.html) function here because, like Java's `assert` keyword, it is disabled unless the `--enable-assertion`/`-ea` flag is passed.)

daira reviewed (approved):
utACK with comment

str4d commented on ui-lib/src/main/java/co/electriccoin/zcash/ui/common/serialization/addressbook/AddressBookKey.kt:
```diff
@@ -0,0 +1,59 @@
+            key.toByteArray(access),
+            salt,
+            null,
+            ADDRESS_BOOK_SALT_SIZE
```
@Milan-Cerovsky this change is incorrect. This field of HKDF is not the length of `salt`, but the output length, i.e. the length of `subKey`.

str4d commented:
[Force-pushed](https://github.com/Electric-Coin-Company/zashi-android/compare/b35f079e323c719928babd37fa0e0af4d120ca0e..20cdaa93ebd07b3bd950a27b272944b139c2d125):

-Rebased on `main`.
- Separated the version encoding spaces. The unencrypted address book format is now versioned separately from the encryption scheme.
- Introduced a file identifier derived from the address book key.
- Fixed the encryption key derivation process.

str4d commented on ui-lib/src/main/java/co/electriccoin/zcash/ui/common/repository/AddressBookRepository.kt:
```diff
@@ -263,6 +283,23 @@ class AddressBookRepositoryImpl(
     }
 
+    private suspend fun getAddressBookKey(): AddressBookKey {
+        val key = addressBookKeyStorageProvider.getAddressBookKey()
```
This will eventually need to be keyed by whatever identifier we use to distinguish between accounts. Currently that is `Account` (being accounts under the same seed phrase); in future it may also need to accommodate imported UFVKs. I could not figure out how to do this, so I'll leave it to @HonzaR or @Milan-Cerovsky to either do this now, or be prepared to migrate the storage provider to enable this later.

Milan-Cerovsky replied:
We will do this as a part of multi-account support for keystone

Milan-Cerovsky replied:
sure why not

daira commented on ui-lib/src/main/java/co/electriccoin/zcash/ui/common/datasource/LocalAddressBookDataSource.kt:
```diff
@@ -59,14 +72,15 @@ class LocalAddressBookDataSourceImpl(
 
     override suspend fun saveContact(
```
Non-blocking:

This will write the address book file on every update to a contact. Are there any situations currently where we update multiple contacts at once? If there are then we should probably have a batch update API.

Milan-Cerovsky replied:
Currently we do not have multiple contacts CRUD operations at once.

Also we do support have parallel CRUD operations - instead we implemented a semaphore and the operations are executed in the order in which they were requested.

daira commented on ui-lib/src/main/java/co/electriccoin/zcash/ui/common/datasource/LocalAddressBookDataSource.kt:
```diff
@@ -104,45 +119,75 @@ class LocalAddressBookDataSourceImpl(
                             .toList(),
                 )
-            writeAddressBookToLocalStorage(addressBook!!)
+            writeAddressBookToLocalStorage(addressBook!!, addressBookKey)
```
Same comment as above.

daira commented on ui-lib/src/main/java/co/electriccoin/zcash/ui/common/datasource/LocalAddressBookDataSource.kt:
```diff
@@ -104,45 +119,75 @@ class LocalAddressBookDataSourceImpl(
                             .toList(),
                 )
-            writeAddressBookToLocalStorage(addressBook!!)
+            writeAddressBookToLocalStorage(addressBook!!, addressBookKey)
```
Same comment as above.

Milan-Cerovsky replied:
done

daira commented on ui-lib/src/main/java/co/electriccoin/zcash/ui/common/datasource/LocalAddressBookDataSource.kt:
```diff
@@ -104,45 +119,75 @@ class LocalAddressBookDataSourceImpl(
+        val encryptedFile = addressBookStorageProvider.getStorageFile(addressBookKey)
+        if (encryptedFile != null) {
+            return try {
+                addressBookProvider.readAddressBookFromFile(encryptedFile, addressBookKey)
```
After this was written, we changed the design so that it should attempt to delete any unencrypted file after it has successfully read, decrypted, and parsed an encrypted file (for account 0, which is all we currently support). Please make this change.

Milan-Cerovsky replied:
done

daira reviewed (approved):
utACK

Milan-Cerovsky commented on ui-lib/src/main/java/co/electriccoin/zcash/ui/common/datasource/LocalAddressBookDataSource.kt:
```diff
@@ -75,21 +89,22 @@ class LocalAddressBookDataSourceImpl(
                             ),
                 )
-            writeAddressBookToLocalStorage(addressBook!!)
+            writeAddressBookToLocalStorage(addressBook!!, addressBookKey)
```
done

HonzaR reviewed (approved):
I just checked that the new logic produces expected result.
